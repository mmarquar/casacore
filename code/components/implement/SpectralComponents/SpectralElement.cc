//# SpectralElement.cc: Describes (a set of related) spectral lines
//# Copyright (C) 2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <trial/Wnbt/SpectralElement.h>

#include <aips/Containers/Record.h>
#include <aips/Containers/RecordInterface.h>
#include <aips/Exceptions/Error.h>
#include <aips/Mathematics/Constants.h>
#include <aips/Mathematics/Math.h>
#include <aips/Quanta/MUString.h>
#include <aips/Utilities/Assert.h>
#include <aips/Utilities/String.h>

#include <aips/iostream.h>

//# Constructors
SpectralElement::SpectralElement() :
  tp_p(SpectralElement::GAUSSIAN), n_p(0),
  par_p(3) {
  par_p(0) = 1.0;
  par_p(1) = 0.0;
  par_p(2) = 2*sqrt(C::ln2)/C::pi;
}

SpectralElement::SpectralElement(SpectralElement::Types tp, const Double ampl,
				 const Double center, const Double sigma) :
  tp_p(tp), n_p(0),
  par_p(3) {
  if (tp != GAUSSIAN) {
    throw(AipsError("SpectralElement: Only GAUSSIAN can have ampl, "
		    "center and sigma"));
  };
  par_p(0) = ampl;
  par_p(1) = center;
  par_p(2) = sigma;
  check();
}

SpectralElement::SpectralElement(const uInt n) :
  tp_p(SpectralElement::POLYNOMIAL), n_p(n),
  par_p(n+1) {
  par_p = 0;
}

SpectralElement::SpectralElement(SpectralElement::Types tp,
				 const Vector<Double> &param) :
  tp_p(tp), n_p(0),
  par_p(0) {
  if (tp_p == GAUSSIAN) {
    if (param.nelements() != 3) {
      throw(AipsError("SpectralElement: GAUSSIAN must have "
		      "3 parameters"));
    };
    par_p.resize(3);
  } else if (tp_p == POLYNOMIAL) {
    if (param.nelements() == 0) {
      throw(AipsError("SpectralElement: POLYNOMIAL must have "
		      "at least 1 parameter"));
    };
    n_p = param.nelements()-1;
    par_p.resize(n_p+1);
  };
  for (uInt i=0; i<param.nelements(); i++) par_p(i) = param(i);
  check();
}

SpectralElement::SpectralElement(const SpectralElement &other) :
  tp_p(other.tp_p), n_p(other.n_p),
  par_p(0) {
  par_p = other.par_p;
  check();
}

SpectralElement::~SpectralElement() {};

SpectralElement &SpectralElement::operator=(const SpectralElement &other) {
  if (this != &other) {
    tp_p = other.tp_p;
    n_p = other.n_p;
    par_p = other.par_p;
    check();
  };
  return *this;
}

Double SpectralElement::operator()(const Double x) const {
  if (tp_p == GAUSSIAN) {
    return  par_p(0)*exp(-(x-par_p(1))*(x-par_p(1))*4*C::ln2/
			 par_p(2)/par_p(2));
  };
  Double s(0);
  if (tp_p == POLYNOMIAL) {
    for (uInt i=n_p; i<=n_p; i--) {
      s += par_p(i); s *= x;
    };
  };    
  return s;
}

const String *const SpectralElement::allTypes(Int &nall,
					      const SpectralElement::Types
					      *&typ) {
  static const String tname[SpectralElement::N_Types] = {
    String("GAUSSIAN"),
    String("POLYNOMIAL") };

  static const SpectralElement::Types oname[SpectralElement::N_Types] = {
    SpectralElement::GAUSSIAN,
    SpectralElement::POLYNOMIAL };

  nall = SpectralElement::N_Types;
  typ    = oname;
  return tname;
}
 
const String &SpectralElement::fromType(SpectralElement::Types tp) {
  Int nall;
  const SpectralElement::Types *typ;
  const String *const tname = SpectralElement::allTypes(nall, typ);
  return tname[tp];
}

Bool SpectralElement::toType(SpectralElement::Types &tp,
			     const String &typname) {
  Int nall;
  const SpectralElement::Types *typ;
  const String *const tname = SpectralElement::allTypes(nall, typ);
  
  // Make sure a value returned
  tp = typ[0];
  Int i = MUString::minimaxNC(typname, SpectralElement::N_Types, tname);
  if (i >= nall) return False;
  tp = typ[i];
  return True;
}

Double SpectralElement::getAmpl() const {
  checkGauss();
  return par_p(0);
};

Double SpectralElement::getCenter() const {
  checkGauss();
  return par_p(1);
};

Double SpectralElement::getSigma() const {
  checkGauss();
  return par_p(2);
};

Double SpectralElement::getFWHM() const {
  checkGauss();
  return sqrt(32.0*C::ln2)*par_p(2);
}

uInt SpectralElement::getDegree() const {
  checkPoly();
  return n_p;
};

void SpectralElement::set(SpectralElement::Types tp,
			  const Vector<Double> &param) {
  tp_p = tp;
  n_p = 0;
  if (tp_p == GAUSSIAN) {
    if (param.nelements() != 3) {
      throw(AipsError("SpectralElement: GAUSSIAN must have "
		      "3 parameters"));
    };
    par_p.resize(3);
  };
  if (tp_p == POLYNOMIAL) {
    if (param.nelements() == 0) {
      throw(AipsError("SpectralElement: POLYNOMIAL must have "
		      "at least 1 parameter"));
    };
    n_p = param.nelements()-1;
    par_p.resize(n_p+1);
  };
  for (uInt i=0; i<param.nelements(); i++) par_p(i) = param(i);
  check();
}

void SpectralElement::setAmpl(Double ampl) {
  checkGauss();
  par_p(0) = ampl;
} 

void SpectralElement::setCenter(Double center) {
  checkGauss();
  par_p(1) = center;
}

void SpectralElement::setSigma(Double sigma) {
  checkGauss();
  par_p(2) = sigma;
  check();
}

void SpectralElement::setDegree(uInt n) {
  checkPoly();
  n_p = n;
  par_p.resize(n_p+1);
  par_p = 0;
}

const String &SpectralElement::ident() const {
  static String myid = "spectrel";
  return myid;
}

void SpectralElement::checkGauss() const {
  if (tp_p != GAUSSIAN) {
    throw(AipsError("SpectralElement: GAUSSIAN element expected"));
  };
}

void SpectralElement::checkPoly() const {
  if (tp_p != POLYNOMIAL) {
    throw(AipsError("SpectralElement: POLYNOMIAL element expected"));
  };
}

void SpectralElement::check() const {
  if (tp_p == GAUSSIAN && par_p(2) <= 0.0) {
    throw(AipsError("SpectralElement: An illegal non-positive sigma was "
		    "specified for a gaussian SpectralElement"));
  };
}

ostream &operator<<(ostream &os, const SpectralElement &elem) {
  switch (elem.getType()) {
    
  case SpectralElement::GAUSSIAN:
  default:
    os << SpectralElement::fromType((elem.getType())) << " element: " << endl;;
    os << "  Amplitude: " << elem.getAmpl() << endl;
    os << "  Center:    " << elem.getCenter() << endl;
    os << "  Sigma:     " << elem.getSigma() << endl;
    break;
  };

  return os;
}
