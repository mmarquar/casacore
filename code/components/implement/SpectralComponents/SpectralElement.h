//# SpectralElement.h: Describes (a set of related) spectral lines
//# Copyright (C) 2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#if !defined(TRIAL_SPECTRALELEMENT_H)
#define TRIAL_SPECTRALELEMENT_H

//# Includes
#include <aips/aips.h>

//# Forward Declarations
class String;
#include <aips/iosfwd.h>

// <summary>
// Describes (a set of related) spectral lines
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="tSpectralFit" demos="">
// </reviewed>

// <prerequisite>
//   <li> <ahlink module=Functionals>Functionals</ahlink> module
// </prerequisite>
//
// <etymology>
// From spectral line and element
// </etymology>
//
// <synopsis>
// The SpectralElement class is a container for a spectral line descriptor.
// It can contain a single line (like a Gaussian profile), or a set of
// related lines (like a doublet or so).
//
// The element can be used in the
// <ahlink class=SpectralFit>SpectralFit</ahlink> class and in the
// <ahlink class=SpectralEstimate>SpectralEstimate</ahlink> class.
//
// The default type is a Gaussian, defined as:
// <srcblock>
//	AMPL.exp[ -(x-CENTER)<sup>2</sup>/2 SIGMA<sup>2</sup>]
// </srcblock>
//
// </synopsis>
//
// <example>
// </example>
//
// <motivation>
// To have a contained fitting of spectral profiles to an observed spectrum
// </motivation>
//
// <todo asof="2001/02/04">
//   <li> add more profile types
// </todo>

class SpectralElement {
 public:

  //# Enumerations
  // Types of spectral lines known
  enum Types {
    // A gaussian profile
    GAUSSIAN,
    // A polynomial baseline
    POLYNOMIAL,
    N_Types
  };

  //# Friends

  //# Constructors
  // Default constructor creates a default Gaussian element with an amplitude
  // of 1; a width (sigma) of 1; a central frequency of zero.
  SpectralElement();
  // Construct with given type and values
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  SpectralElement(SpectralElement::Types tp, const Double ampl,
		  const Double center, const Double sigma);
  // Construct an n-degree polynomial
  explicit SpectralElement(const uInt n);
  // Copy constructor (deep copy)
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  SpectralElement(const SpectralElement &other);

  //# Operators
  // Assignment (copy semantics)
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  SpectralElement &operator=(const SpectralElement &other);

  //# Member functions
  // Get all the types available as String and codes, and number available
  static const String *const allTypes(Int &nall,
				      const SpectralElement::Types *&typ);
  // Get a string from the type
  static const String &fromType(SpectralElement::Types tp);
  // Get a type from a (non-case sensitive; minimum match) String
  static Bool toType(SpectralElement::Types &tp,
		     const String &typName);

  // Get the data for this element
  // <group>
  // Get type of this element
  SpectralElement::Types getType() const { return tp_p; }
  // Get amplitude
  Double getAmpl() const { return ampl_p; }
  // Get center value
  Double getCenter() const { return center_p; }
  // Get the width
  // <group>
  Double getSigma() const { return sigma_p; }
  Double getFWHM() const;
  // </group>
  // Get the degree of e.g. polynomial
  uInt getDegree() const { return n_p; }
  // </group>

  // Set data for element
  // <group>
  // Set all data
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  void set(SpectralElement::Types tp, const Double ampl,
	   const Double center, const Double sigma);
  // Set type
  void setType(SpectralElement::Types tp);
  // Set amplitude
  void setAmpl(Double ampl);
  // Set center
  void setCenter(Double center);
  // Set width
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  void setSigma(Double sigma);
  // </group>
  // Set degree
  void setDegree(uInt n);

  //#Destructor
  // Destructor
  ~SpectralElement();

 private:
  //#Data
  // type of element
  SpectralElement::Types tp_p;
  // A number (like polynomial degree or number of doublet lines
  uInt n_p;
  // Amplitude
  Double ampl_p;
  // Center frequency
  Double center_p;
  // Width
  Double sigma_p;

  //# Member functions
  // Check for valid values: sigma should not be zero.
  // <thrown>
  //   <li> AipsError if sigma == 0.0
  // </thrown>
  void SpectralElement::check();

};

//# Global functions
// <summary> Global functions </summary>
// <group name=Output>
// Output declaration
ostream &operator<<(ostream &os, const SpectralElement &elem);
// </group>

#endif


