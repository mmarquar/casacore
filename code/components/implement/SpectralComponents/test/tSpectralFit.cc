//# tSpectralFit.cc -- test SpectralElement; SpectralEstimate and SpectralFit
//# Copyright (C) 2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <aips/aips.h>
#include <trial/Wnbt/SpectralElement.h>
#include <trial/Wnbt/SpectralEstimate.h>
#include <trial/Wnbt/SpectralFit.h>
#include <trial/Tasking/PGPlotter.h>
#include <aips/Exceptions/Error.h>
#include <aips/Arrays/Vector.h>
#include <aips/Mathematics/Random.h>
#include <aips/Inputs/Input.h>
#include <aips/iostream.h>

int main(int argc, char **argv) {
      
  // Inputs
  Input inputs(1);
  inputs.version("$Id$");
  inputs.create("plotter", "", "plotter"); 
  inputs.readArguments(argc, argv);
  const String device = inputs.getString("plotter"); 

  PGPlotter plotter(device, 3, 3);
  // get estimates
  {
    cout << "Test SpectralEstimate" << endl;
    cout << "---------------------------------------------------" << endl;

    const Int NPAR = 9;
    const Int NPTS = 100;
    uInt q = 2;
    Int r;
    Int n = NPTS;
    Int np = NPAR;
    Double rms = 0.5;
    Double cutoff= 1.0;
    Double minsig = 0.5;
    Double par[NPAR];
    Vector<Float> y(NPTS);
    Vector<Float> xy(NPTS);

    par[0] = 5.0; par[1] = 45.0; par[2] = 1.1;
    par[3] = 4.0; par[4] = 50.0; par[5] = 1.1;
    par[6] = 6.0; par[7] = 55.0; par[8] = 1.1;
    for (Int i = 0; i<n; i++) {
      y(i) = 0.0;
      xy(i) = i;
      for (Int j = 0; j < np; j += 3) {
	if (par[j] != 0.0) {
	  Double a = par[j];
	  Double c = (par[j+1] - Double(i));
	  Double s = par[j+2];
	  y(i) += a * exp( -0.5 * c / s * c / s );
	};
      };
    };
    cout << "Made spectrum with components: " << endl;
    for (uInt i=0; i<3; i++) {
      cout << SpectralElement(SpectralElement::GAUSSIAN,
			      par[3*i+0], par[3*i+1], par[3*i+2]) << endl;
    };
    plotter.env(xy(0), xy(n-1), -1, 12, 0, 0);
    plotter.lab("", "", "Noiseless data with 3 components and "
		"estimate residuals");
    plotter.line(xy, y);
    {
      SpectralEstimate est(rms, cutoff, minsig);
      Vector<Float> deriv(n);
      r = est.estimate(y, &deriv);
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << est.element(i) << endl;
      };
      Vector<Float> res(NPTS);
      Float rmn = 1e6;
      Float rmx = -1e6;
      for (Int i=0; i<n; i++) {
	res(i) = y(i);
	for (Int j = 0; j < r; j++) {
	  Double a = est.element(j).getAmpl();
	  Double c = (est.element(j).getCenter() - Double(i));
	  Double s = est.element(j).getSigma();
	  res(i) -= a * exp( -log(16.0) * c / s * c / s );
	  rmn = min(rmn, res(i));
	  rmx = max(rmx, res(i));
	};
      };
      plotter.sci(2);
      plotter.line(xy, res);
      plotter.env(xy(0), xy(n-1), rmn, rmx, 0, 0);
      plotter.sci(1);
      plotter.lab("", "", "Noiseless spectrum with 2nd derivative");
      plotter.line(xy, y);
      plotter.sci(2);
      plotter.line(xy, deriv);
      plotter.env(xy(0), xy(n-1), -1, 12, 0, 0);
      plotter.sci(1);
      plotter.lab("", "", "Noiseless spectrum with estimates");
      plotter.line(xy, y);
      plotter.sci(2);
      for (Int j=0; j<r; j++) {
	for (Int i = 0; i<n; i++) {
	  Double a = est.element(j).getAmpl();
	  Double c = (est.element(j).getCenter() - Double(i));
	  Double s = est.element(j).getSigma();
	  res(i) = a * exp( -log(16.0) * c / s * c / s );
	};
	plotter.line(xy, res);
      };
    }
    {
      SpectralEstimate est(rms, cutoff, minsig);
      est.setWindowing(True);
      r = est.estimate(y);
      cout << "Window (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << est.element(i) << endl;
      };
    }
    {
      q = 1;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      r = est.estimate(y);
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << est.element(i) << endl;
      };
    }
    {
      q = 1;
      rms = cutoff = minsig = 0.0;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      r = est.estimate(y);
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << est.element(i) << endl;
      };
    }
    {
      q = 2;
      rms = cutoff = minsig = 0.0;
      SpectralEstimate est(rms, cutoff, minsig);
      est.setQ(q);
      r = est.estimate(y);
      cout << "Found (using rms, cutoff, minsig, q): (" <<
	rms << ", " << cutoff << ", " << minsig << ", " << q << ") " <<
	r << " estimates" << endl;
      for (Int i=0; i<r; i++) {
	cout << "Estimate " << i << ": " << est.element(i) << endl;
      };
    }
  }

  // test fitter
  {
    // Number of components
    const uInt ncomp = 4;
    // Component data
    Double ampl[ncomp] = { 1, 2.5, 0.5, 5 };
    Double center[ncomp] = { 0.25, 0.5, 0.75, 0.3 };
    Double sigma[ncomp] = { 2, 4, 8, 3 };
    MLCG genit;
    Normal noise(&genit, 0.0, 0.1);
    Vector<Double> freq(1024);
    Vector<Float> ffreq(1024);
    for (uInt i=0; i<1024; i++) {
      freq(i) = 1400 + i*10.0/1024.0;
      ffreq(i) = freq(i);
    };

    try {
      cout << "Test SpectralFit" << endl;
      cout << "---------------------------------------------------" << endl;
      
      SpectralElement el[ncomp] = {
	SpectralElement(SpectralElement::GAUSSIAN, ampl[0],
			center[0]*10.+1400., sigma[0]*10./1024.),
	SpectralElement(SpectralElement::GAUSSIAN, ampl[1],
			center[1]*10.+1400., sigma[1]*10./1024.),
	SpectralElement(SpectralElement::GAUSSIAN, ampl[2],
			center[2]*10.+1400., sigma[2]*10./1024.),
	SpectralElement(SpectralElement::GAUSSIAN, ampl[3],
			center[3]*10.+1400., sigma[3]*10./1024.) };
      
      cout << "Spectral elements: " << endl;
      for (uInt j=0; j<ncomp; j++) cout << el[j] << endl;
      cout << "---------------------------------------------------" << endl;
      
      Vector<Double> dat(1024);
      Vector<Float> fdat(1024);
      for (uInt i=0; i<1024; i++) {
	dat(i) = 0;
	for (uInt j=0; j<ncomp; j++) {
	  dat(i) += el[j].getAmpl()*exp(-(freq(i)-el[j].getCenter())*
					(freq(i)-el[j].getCenter())*log(16.0)/
					el[j].getSigma()/
					el[j].getSigma());
	};
	dat(i) += noise();
	fdat(i) = dat(i);
      };
      cout << "Data for frequencies " << freq(300) << " - " << freq(315) <<
	endl;
      for (uInt i=300; i<316; i++) cout << freq(i) << ": " << dat(i) << endl;
      cout << "---------------------------------------------------" << endl;
      
      plotter.env(ffreq(0), ffreq(1023), -1, 12, 0, 0);
      plotter.sci(1);
      plotter.lab("", "", "Synthetic spectra and residual after fit");
      plotter.line(ffreq, fdat);

      cout << "Specify fitter on the 4 gaussians:" << endl;
      SpectralFit fitter;
      for (uInt i=0; i<ncomp; i++) fitter.addFitElement(el[i]);
      for (uInt i=0; i<fitter.getNElements(); i++) {
	cout << fitter.getElement(i) << endl;
      };
      cout << "---------------------------------------------------" << endl;
      cout << "Execute the fitter, can do (1:Y): " <<
	fitter.fit(freq, dat) << endl;
      cout << "The results: " << endl;
      for (uInt i=0; i<fitter.getNElements(); i++) {
	cout << fitter.getElement(i) << endl;
      };
      cout << "---------------------------------------------------" << endl;
      
      cout << "Different start values: " << endl;
      for (uInt i=0; i<ncomp; i++) {
	el[i].setAmpl(0.9*el[i].getAmpl());
	el[i].setSigma(0.95*el[i].getSigma());
	fitter.setFitElement(i, el[i]);
      };
      for (uInt i=0; i<fitter.getNElements(); i++) {
	cout << fitter.getElement(i) << endl;
      };
      cout << "Execute the fitter, can do (1:Y): " <<
	fitter.fit(freq, dat) << endl;
      cout << "The results: " << endl;
      for (uInt i=0; i<fitter.getNElements(); i++) {
	cout << fitter.getElement(i) << endl;
      };
      
      cout << "---------------------------------------------------" << endl;
      cout << "Differences: " << endl;
      Vector<Double> xdat(1024);
      Vector<Float> fxdat(1024);
      Double mx(-1e6);
      Double mn(1e6);
      Double avg(0);
      Double sg(0);
      for (uInt i=0; i<1024; i++) {
	xdat(i) = dat(i);
	for (uInt j=0; j<fitter.getNElements(); j++) {
	  xdat(i) -= fitter.getElement(j).getAmpl()*
	    exp(-(freq(i)-fitter.getElement(j).getCenter())*
		(freq(i)-fitter.getElement(j).getCenter())*log(16.0)/
		fitter.getElement(j).getSigma()/
		fitter.getElement(j).getSigma());
	  fxdat(i) = xdat(i);
	};
	mx = (xdat(i)>mx) ? xdat(i) : mx;
	mn = (xdat(i)<mn) ? xdat(i) : mn;
	avg += xdat(i);
	sg += xdat(i)*xdat(i);
      };
      avg /= 1024.;
      sg = sqrt(sg/1024./1023.);
      cout << "Min difference: " << mn <<
	", max: " << mx << ", average: " << avg <<
	", sigma: " << sg << endl;
      plotter.sci(2);
      plotter.line(ffreq, fxdat);
      cout << "---------------------------------------------------" << endl;
      
      cout << "---------------------------------------------------" << endl;
      cout << "Estimates: " << endl;
      SpectralEstimate mest(0.1, 0.5);
      mest.setQ(5);
      Int mr = mest.estimate(fdat);
      cout << "Found " << mr << " estimates" << endl;
      for (Int i=0; i<mr; i++) {
	cout << "Estimate " << i << ": " << mest.element(i) << endl;
      };
      plotter.env(ffreq(0), ffreq(1023), -1, 12, 0, 0);
      plotter.sci(1);
      plotter.lab("", "", "Synthetic spectra and residual after estimates");
      plotter.line(ffreq, fdat);
      for (uInt i=0; i<1024; i++) {
	xdat(i) = 0;
	for (Int j = 0; j < mr; j++) {
	  Double a = mest.element(j).getAmpl();
	  Double c = (mest.element(j).getCenter() - Double(i));
	  Double s = mest.element(j).getSigma();
	  xdat(i) += a * exp( -log(16.0) * c / s * c / s );
	};
	mx = (xdat(i)>mx) ? xdat(i) : mx;
	mn = (xdat(i)<mn) ? xdat(i) : mn;
      };
      plotter.sci(2);
      plotter.line(ffreq, fxdat);
      plotter.env(ffreq(0), ffreq(1023), -1, 12, 0, 0);
      plotter.sci(1);
      plotter.lab("", "", "Synthetic spectrum with estimates");
      plotter.line(ffreq, fdat);
      plotter.sci(2);
      for (Int j=0; j<mr; j++) {
	for (Int i = 0; i<1024; i++) {
	  Double a = mest.element(j).getAmpl();
	  Double c = (mest.element(j).getCenter() - Double(i));
	  Double s = mest.element(j).getSigma();
	  fxdat(i) = a * exp( -log(16.0) * c / s * c / s );
	};
	plotter.line(ffreq, fxdat);
      };
      plotter.env(ffreq(0), ffreq(1023), -1, 12, 0, 0);
      
      cout << "---------------------------------------------------" << endl;
      
    } catch (AipsError x) {
      cout << x.getMesg() << endl;
    };
  }
  
  exit(0);
}
