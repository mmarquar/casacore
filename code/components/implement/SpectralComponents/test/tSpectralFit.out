Test SpectralFit
---------------------------------------------------
Spectral elements: 
GAUSSIAN element: 
  Amplitude: 1
  Center:    1402.5
  Sigma:     0.0195312

GAUSSIAN element: 
  Amplitude: 2.5
  Center:    1405
  Sigma:     0.0390625

GAUSSIAN element: 
  Amplitude: 0.5
  Center:    1407.5
  Sigma:     0.078125

GAUSSIAN element: 
  Amplitude: 5
  Center:    1403
  Sigma:     0.0292969

---------------------------------------------------
Data for frequencies 1402.93 - 1403.08
1402.93: 0.346479
1402.94: -0.759988
1402.95: -0.445616
1402.96: 0.0542814
1402.97: 0.00845671
1402.98: 1.01413
1402.99: 3.10099
1403: 4.92518
1403.01: 4.78433
1403.02: 1.37155
1403.03: 0.132033
1403.04: -0.839767
1403.05: 0.0682422
1403.06: -0.0821468
1403.07: -1.18733
1403.08: 0.789914
---------------------------------------------------
Specify fitter on the 4 gaussians:
GAUSSIAN element: 
  Amplitude: 1
  Center:    1402.5
  Sigma:     0.0195312

GAUSSIAN element: 
  Amplitude: 2.5
  Center:    1405
  Sigma:     0.0390625

GAUSSIAN element: 
  Amplitude: 0.5
  Center:    1407.5
  Sigma:     0.078125

GAUSSIAN element: 
  Amplitude: 5
  Center:    1403
  Sigma:     0.0292969

---------------------------------------------------
Execute the fitter, can do (1:Y): 1
The results: 
GAUSSIAN element: 
  Amplitude: 0.758448
  Center:    1402.5
  Sigma:     0.0242276

GAUSSIAN element: 
  Amplitude: 2.91229
  Center:    1405
  Sigma:     0.0342584

GAUSSIAN element: 
  Amplitude: 0.558048
  Center:    1407.43
  Sigma:     0.000298385

GAUSSIAN element: 
  Amplitude: 5.39508
  Center:    1403
  Sigma:     0.0263626

---------------------------------------------------
Different start values: 
GAUSSIAN element: 
  Amplitude: 0.9
  Center:    1402.5
  Sigma:     0.0185547

GAUSSIAN element: 
  Amplitude: 2.25
  Center:    1405
  Sigma:     0.0371094

GAUSSIAN element: 
  Amplitude: 0.45
  Center:    1407.5
  Sigma:     0.0742188

GAUSSIAN element: 
  Amplitude: 4.5
  Center:    1403
  Sigma:     0.027832

Execute the fitter, can do (1:Y): 1
The results: 
GAUSSIAN element: 
  Amplitude: 0.725073
  Center:    1402.5
  Sigma:     0.0245704

GAUSSIAN element: 
  Amplitude: 2.91207
  Center:    1405
  Sigma:     0.0342639

GAUSSIAN element: 
  Amplitude: 0.562647
  Center:    1407.43
  Sigma:     0.0258912

GAUSSIAN element: 
  Amplitude: 5.39369
  Center:    1403
  Sigma:     0.0263769

---------------------------------------------------
Differences: 
Min difference: -1.73915, max: 1.60124, average: 0.0173243, sigma: 0.0159238
---------------------------------------------------
